salt/_modules/metalk8s.py

```diff
@@ -355,3 +358,164 @@ def cmp_sorted(*args, **kwargs):
+        )):
+            # Don't remove if it exists in file_roots (any saltenv)
+            all_roots = itertools.chain.from_iterable(
+                __opts__["file_roots"].values()
+            )
+            in_roots = any(sfn.startswith(root) for root in all_roots)
+            # Only clean up files that exist
+            if os.path.exists(sfn) and not in_roots:
+                os.remove(sfn)
+
+    target_dir = os.path.dirname(name)
+    target_exists = os.path.isfile(name) or os.path.islink(name)
+    hash_type = source_sum.get("hash_type", __opts__["hash_type"])
+
+    if not source:
+        return _error(ret, "Must provide a source")
+    if not os.path.isdir(target_dir):
+        return _error(
+            ret, "Target directory {} does not exist".format(target_dir)
+        )
+
+    if source_filename:
+        # File should be already cached, verify its checksum
+        cached_hash = get_hash(source_filename, form=hash_type)
+        if source_sum.get("hsum") != cached_hash:
+            log.debug(
+                "Cached source file {} does not match expected checksum, "
+                "will fetch it again".format(source_filename)
+            )
+            source_filename = ''  # Reset source filename to fetch it again
+
+    if not source_filename:
+        # File is not present or outdated, cache it
+        source_filename = __salt__["cp.cache_file"](source, saltenv)
+        if not source_filename:
+            return _error(
+                ret, "Source file '{}' not found".format(source)
+            )
+
+        # Recalculate source sum now that file has been cached
+        source_sum = {
+            "hash_type": hash_type,
+            "hsum": get_hash(source_filename, form=hash_type)
+        }
+
+    # Check changes if the target file exists
+    if target_exists:
+        if os.path.islink(name):
+            real_name = os.path.realpath(name)
+        else:
+            real_name = name
+
+        target_hash = get_hash(real_name, hash_type)
+        source_hash = source_sum.get("hsum")
+
+        # Check if file needs to be replaced
+        if source_hash != target_hash:
+            # Print a diff equivalent to diff -u old new
+            if __salt__["config.option"]("obfuscate_templates"):
+                ret["changes"]["diff"] = "<Obfuscated Template>"
+            else:
+                try:
+                    ret["changes"]["diff"] = __salt__["file.get_diff"](
+                        real_name, source_filename, show_filenames=False
+                    )
+                except CommandExecutionError as exc:
+                    ret["changes"]["diff"] = exc.strerror
+
+    else:  # target file does not exist
+        ret["changes"]["diff"] = "New file"
+        real_name = name
+
+    if ret["changes"] and not __opts__["test"]:
+        # The file needs to be replaced
+        try:
+            _atomic_copy(source_filename, real_name)
```

Comment:
Wouldn't it make sense for `_atomic_copy` *not* to have an arbitrary default for `tmp_prefix` and instead pass one explicitly from here?

Reply:
Sure, if we want it to be less specific about our use-case. I'll make this change.